notebooks/COS/ExceptionReport/cos_functions.py

```diff
@@ -0,0 +1,434 @@
+                        print(
+                            f"In range on {i}-th segment with limits:",
+                            min(wvln_),
+                            max(wvln_),
+                            f"\nUnweighted mean SNR over the range {snr_range} is: {weight_avg_snr}",
+                        )
+
+                if weighted:
+                    weight_avg_snr = np.average(
+                        np.sqrt(gcount_range),
+                        weights=np.nan_to_num(gcount_range / gross_range, nan=1e-30),
+                    )
+                    if verbose:
+                        print(
+                            f"In range on {i}-th segment with limits:",
+                            min(wvln_),
+                            max(wvln_),
+                            f"\nEXPTIME weighted average SNR over the range {snr_range} is: {weight_avg_snr}",
+                        )
+            else:
+                snr_array.append([-1, -1, -1])
+                if verbose:
+                    print(
+                        f"Out of range on {i}-th segment with limits:",
+                        min(wvln_),
+                        max(wvln_),
+                    )
+    if (  # Check whether the code has found any specified wavelength ranges
+        all([elem == -1 for elem in (snr_array[0][2], snr_array[1][2])])
+    ) & (  # Check that the user specified wavelength ranges
+        snr_range != [-1, -1]
+    ):
+        if verbose:  # Then print to the user that the ranges weren't found
+            print("\nThe input range was not found in any segment!\n")
+    if segsFound > 1:
+        if verbose:
+            print(
+                "\nThis range was found on multiple segments, (grating = G230L?) ,which at present is not fully supported. The returned array should be accurate, but the mean may be incorrect."
+            )
+    return weight_avg_snr, snr_array
+
+
+# %%
+def withinPercent(val1, val2, percent=1.0):
+    """
+    Primarily created for testing, this function evaluates whether two values are 'close-enough' to one another, i.e. within a percent value, that they could only differ by slight pipeline changes; This one is defined such that at close values, the percent difference is accurate.
+
+    Parameters:
+    val1, val2 (numerical) : Values to compare.
+    percent (float) : Returns value of true if the values are within this percent.
+
+    Returns:
+    bool : Whether or not values are within the specified percent.
+    float : Percent they are off by.
+    """
+    if (val1 == np.nan) | (val2 == np.nan):
+        print("One of your values is NOT A NUMBER")
+    lowval = np.min(np.array([val1, val2]))
+    absDif = np.abs(np.subtract(val1, val2))
+    percentDif = np.abs(100 * (absDif / lowval))
+    within_percent_bool = percentDif <= percent
+    return within_percent_bool, percentDif
+
+
+# %%
+"""
+This section is for testing and validating our S/N Calculations.
+"""
+if __name__ == "__main__":
```

Comment:
I was reviewing the script and noticed the `if __name__ == "__main__":` block used for testing. I just wanted to flag that this might not be immediately clear to many users, especially those working primarily in notebooks. Typically, notebook users won't be running scripts directly from the command line, and this main block might not be the best fit for those workflows. 
- If you want users to be able to run this test from their notebooks, it might be helpful to provide some instructions on how to do so. Notebook users may not know how to interact with the main block, and it could create confusion. OR
- You could move the logic in the main block into a dedicated function. This way, users can call the function directly from their notebook or another script without needing to figure out how to trigger the main() block. It will also allow you to run tests with a simple function call rather than hardcoding things.
-  also noticed that the file path in the main() block is hardcoded, but it seems like the file doesn’t exist. It might be better to either check for the file’s existence or make the path dynamic (e.g., allow the user to input the file path).
In short :)
If testing isn’t needed in the notebook workflow, consider removing the `if __name__ == "__main__":` block entirely.
If you want users to be able to test from the notebook, encapsulate the logic into a separate function with clear instructions.
Let me know what you think, and feel free to reach out if you'd like to discuss this further. Thanks again for all your hard work on this!